dask/dataframe/core.py

```diff
@@ -3570,6 +3573,10 @@ def __setitem__(self, key, value):
         elif isinstance(key, pd.Index) and not isinstance(value, DataFrame):
             key = list(key)
             df = self.assign(**{k: value for k in key})
+        elif isinstance(key, (pd.DataFrame, DataFrame)):
+            df = self.where(~key, value)
+        elif not isinstance(key, str):
+            raise NotImplementedError(f"Item assignment with {type(key)} not supported")
```

Comment:
Perhaps a test for this? 
If pandas doesn't support it either, this can be a ValueError instead

Reply:
or a column that has a name that is not a string

Reply:
I added a test :)